File: CPP/negocio/JuegoDefensaTorres.cpp
#ifndef JUEGO_DEFENSA_TORRES_CPP
#define JUEGO_DEFENSA_TORRES_CPP

#include <vector>
#include <queue>
#include <stack>
#include <algorithm>
#include "../modelo/Torre.cpp"
#include "../modelo/Cozy.cpp"
#include "../modelo/Oleada.cpp"
#include "../modelo/AccionColocarTorre.cpp"

class JuegoDefensaTorres {
private:
    std::vector<Torre*> torresColocadas;
    std::stack<AccionColocarTorre> historialDeshacer;
    std::stack<AccionColocarTorre> historialRehacer;

    std::queue<Oleada> oleadasPendientes;
    std::queue<Cozy> colaRuta;

    int longitudRuta;
    int vidaJugador;
    int puntuacion;
    int oleadaActual;

public:
    JuegoDefensaTorres(int longitudRuta, int vidaInicialJugador) {
        this->longitudRuta = longitudRuta;
        this->vidaJugador = vidaInicialJugador;
        this->puntuacion = 0;
        this->oleadaActual = 0;
    }

    void colocarTorre(Torre* torre) {
        torresColocadas.push_back(torre);
        int indice = (int)torresColocadas.size() - 1;
        historialDeshacer.push(AccionColocarTorre(torre, indice));
        while (!historialRehacer.empty()) {
            historialRehacer.pop();
        }
    }

    bool deshacerUltimaColocacion() {
        if (historialDeshacer.empty()) {
            return false;
        }
        AccionColocarTorre accion = historialDeshacer.top();
        historialDeshacer.pop();
        auto it = std::find(torresColocadas.begin(), torresColocadas.end(), accion.getTorre());
        if (it != torresColocadas.end()) {
            torresColocadas.erase(it);
        }
        historialRehacer.push(accion);
        return true;
    }

    bool rehacerColocacion() {
        if (historialRehacer.empty()) {
            return false;
        }
        AccionColocarTorre accion = historialRehacer.top();
        historialRehacer.pop();
        int indice = std::min(accion.getIndiceEnLista(), (int)torresColocadas.size());
        torresColocadas.insert(torresColocadas.begin() + indice, accion.getTorre());
        historialDeshacer.push(accion);
        return true;
    }

    void agregarOleada(Oleada oleada) {
        oleadasPendientes.push(oleada);
    }

    bool iniciarSiguienteOleada() {
        if (oleadasPendientes.empty()) {
            return false;
        }
        Oleada oleada = oleadasPendientes.front();
        oleadasPendientes.pop();
        this->oleadaActual = oleada.getNumero();
        for (int i = 1; i <= oleada.getCantidadEnemigos(); i++) {
            colaRuta.push(Cozy("Cozy-O" + std::to_string(oleada.getNumero()) + "-" + std::to_string(i),
                                oleada.getPvPorEnemigo()));
        }
        return true;
    }

    void avanzarQuantum() {
        int enemigosEnEsteQuantum = (int)colaRuta.size();
        for (int i = 0; i < enemigosEnEsteQuantum; i++) {
            if (colaRuta.empty()) {
                break;
            }
            Cozy cozy = colaRuta.front();
            colaRuta.pop();
            for (Torre* torre : torresColocadas) {
                if (torre->enRango(cozy.getPosicionEnRuta())) {
                    cozy.recibirDano(torre->calcularDano());
                }
            }
            if (!cozy.estaVivo()) {
                puntuacion += 10;
                continue;
            }
            cozy.avanzar();
            if (cozy.alcanzoElFinal(longitudRuta)) {
                vidaJugador--;
                continue;
            }
            colaRuta.push(cozy);
        }
    }

    bool oleadaEnCurso() const {
        return !colaRuta.empty();
    }

    const std::vector<Torre*>& getTorresColocadas() const { return torresColocadas; }
    int getVidaJugador() const { return vidaJugador; }
    int getPuntuacion() const { return puntuacion; }
    int getOleadaActual() const { return oleadaActual; }
    int getCantidadEnColaRuta() const { return (int)colaRuta.size(); }
    int getPilaDeshacerTamano() const { return (int)historialDeshacer.size(); }
    int getPilaRehacerTamano() const { return (int)historialRehacer.size(); }
};

#endif

File: Java/src/negocio/JuegoDefensaTorres.java
package negocio;

import modelo.AccionColocarTorre;
import modelo.Cozy;
import modelo.Oleada;
import modelo.Torre;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Stack;

public class JuegoDefensaTorres {

    private final List<Torre> torresColocadas = new ArrayList<>();
    private final Stack<AccionColocarTorre> historialDeshacer = new Stack<>();
    private final Stack<AccionColocarTorre> historialRehacer = new Stack<>();

    private final Queue<Oleada> oleadasPendientes = new LinkedList<>();
    private final Queue<Cozy> colaRuta = new LinkedList<>();

    private final int longitudRuta;
    private int vidaJugador;
    private int puntuacion;
    private int oleadaActual;

    public JuegoDefensaTorres(int longitudRuta, int vidaInicialJugador) {
        this.longitudRuta = longitudRuta;
        this.vidaJugador = vidaInicialJugador;
        this.puntuacion = 0;
        this.oleadaActual = 0;
    }

    public void colocarTorre(Torre torre) {
        torresColocadas.add(torre);
        int indice = torresColocadas.size() - 1;
        historialDeshacer.push(new AccionColocarTorre(torre, indice));
        historialRehacer.clear();
    }

    public boolean deshacerUltimaColocacion() {
        if (historialDeshacer.isEmpty()) {
            return false;
        }
        AccionColocarTorre accion = historialDeshacer.pop();
        torresColocadas.remove(accion.getTorre());
        historialRehacer.push(accion);
        return true;
    }

    public boolean rehacerColocacion() {
        if (historialRehacer.isEmpty()) {
            return false;
        }
        AccionColocarTorre accion = historialRehacer.pop();
        int indice = Math.min(accion.getIndiceEnLista(), torresColocadas.size());
        torresColocadas.add(indice, accion.getTorre());
        historialDeshacer.push(accion);
        return true;
    }

    public void agregarOleada(Oleada oleada) {
        oleadasPendientes.add(oleada);
    }

    public boolean iniciarSiguienteOleada() {
        Oleada oleada = oleadasPendientes.poll();
        if (oleada == null) {
            return false;
        }
        this.oleadaActual = oleada.getNumero();
        for (int i = 1; i <= oleada.getCantidadEnemigos(); i++) {
            colaRuta.add(new Cozy("Cozy-O" + oleada.getNumero() + "-" + i, oleada.getPvPorEnemigo()));
        }
        return true;
    }

    public void avanzarQuantum() {
        int enemigosEnEsteQuantum = colaRuta.size();
        for (int i = 0; i < enemigosEnEsteQuantum; i++) {
            Cozy cozy = colaRuta.poll();
            if (cozy == null) {
                break;
            }
            for (Torre torre : torresColocadas) {
                if (torre.enRango(cozy.getPosicionEnRuta())) {
                    cozy.recibirDano(torre.calcularDano());
                }
            }
            if (!cozy.estaVivo()) {
                puntuacion += 10;
                continue;
            }
            cozy.avanzar();
            if (cozy.alcanzoElFinal(longitudRuta)) {
                vidaJugador--;
                continue;
            }
            colaRuta.add(cozy);
        }
    }

    public boolean oleadaEnCurso() {
        return !colaRuta.isEmpty();
    }

    public List<Torre> getTorresColocadas() { return torresColocadas; }
    public int getVidaJugador() { return vidaJugador; }
    public int getPuntuacion() { return puntuacion; }
    public int getOleadaActual() { return oleadaActual; }
    public int getCantidadEnColaRuta() { return colaRuta.size(); }
    public int getPilaDeshacerTamano() { return historialDeshacer.size(); }
    public int getPilaRehacerTamano() { return historialRehacer.size(); }
}
